File: Sources/CodexIsland/TaskSideChatTitle.swift
import Foundation

/// Projects only the side tab's label; the prompt and its context are discarded.
/// Desktop 26.908 names a side tab after its first submitted prompt, separately
/// from the thread title. A later turn or a generated title is not a substitute.
enum TaskSideChatTitle {
    static func from(_ state: [String: Any]) -> String? {
        guard state["sideConversation"] as? Bool == true
                || state["sideConversationParentNavigationPath"] as? String != nil,
              let turn = firstTurn(in: state),
              let params = turn["params"] as? [String: Any],
              let input = params["input"] as? [[String: Any]],
              let text = input.first(where: { $0["type"] as? String == "text" })?["text"] as? String else {
            return nil
        }
        // Desktop removes its prepended context at the last request marker.
        // Bound work and retain only a small label, never the surrounding input.
        guard text.utf8.count <= 1_048_576 else { return nil }
        let expression = try? NSRegularExpression(pattern: "## My request(?: for Codex)?:")
        let range = NSRange(text.startIndex..., in: text)
        let lastMarker = expression?.matches(in: text, range: range).last
        let start = lastMarker.map { NSMaxRange($0.range) } ?? 0
        let prompt = (text as NSString).substring(from: start).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !prompt.isEmpty, prompt.utf8.count <= 8_192 else { return nil }

        // Derive the plain-text Markdown label without loading HTML or remote
        // resources. Exact AX label matching remains necessary: Foundation and
        // Desktop can render unusual Markdown/HTML differently. Angle brackets
        // in inline/fenced code (for example C++ templates) are valid input.
        guard let markdown = try? AttributedString(markdown: prompt,
                                                   options: .init(interpretedSyntax: .full)) else { return nil }
        // Full Markdown parsing represents paragraph boundaries as intents.
        // Insert separators between blocks before collapsing whitespace.
        var plain = ""
        var previousIntent: PresentationIntent?
        for run in markdown.runs {
            if !plain.isEmpty, run.presentationIntent != previousIntent { plain += " " }
            plain += String(markdown[run.range].characters)
            previousIntent = run.presentationIntent
        }
        let title = plain.split(whereSeparator: { $0.isWhitespace }).joined(separator: " ")
        return title.isEmpty ? nil : title
    }

    private static func firstTurn(in state: [String: Any]) -> [String: Any]? {
        if let historyState = state["turnHistory"] as? [String: Any],
           historyState["kind"] as? String == "canonical" {
            guard let history = historyState["history"] as? [String: Any],
                  let firstIsland = (history["islands"] as? [[String: Any]])?.first,
                  let boundary = firstIsland["olderBoundary"] as? [String: Any],
                  boundary["status"] as? String == "exhausted",
                  let key = (firstIsland["entries"] as? [[String: Any]])?.first?["value"] as? String,
                  let entities = history["entitiesByKey"] as? [String: Any] else { return nil }
            return entities[key] as? [String: Any]
        }
        if let pagination = state["turnsPagination"] as? [String: Any],
           pagination["hasLoadedOldest"] as? Bool == false { return nil }
        return (state["turns"] as? [[String: Any]])?.first
    }
}

File: windows/CodexIsland.Core/TaskSideChatTitle.cs
using System.Net;
using System.Text;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace CodexIsland.Core;

/// <summary>Keep only a small plain text tab label, never the prompt context or HTML resources.</summary>
public static class TaskSideChatTitle
{
    private static readonly TimeSpan RegexTimeout = TimeSpan.FromMilliseconds(100);
    public static string? From(JsonObject state)
    {
        if (!state["sideConversation"].Boolean() && state["sideConversationParentNavigationPath"].Text() is null) return null;
        var turn = FirstTurn(state);
        var text = (turn?["params"] as JsonObject)?["input"].Objects().FirstOrDefault(item => item["type"].Text() == "text")?["text"].Text();
        if (text is null || Encoding.UTF8.GetByteCount(text) > 1_048_576) return null;
        try
        {
            var markers = Regex.Matches(text, "## My request(?: for Codex)?:", RegexOptions.None, RegexTimeout);
            var prompt = (markers.Count > 0 ? text[(markers[^1].Index + markers[^1].Length)..] : text).Trim();
            if (prompt.Length == 0 || Encoding.UTF8.GetByteCount(prompt) > 8_192) return null;
            // A conservative projection supports ordinary Markdown. The navigator must require an
            // exact accessible label; unsupported syntax fails closed rather than selecting a nearby tab.
            prompt = Regex.Replace(prompt, @"(?m)^\s*```[^\r\n]*\r?\n|(?m)^\s*```\s*$", "", RegexOptions.None, RegexTimeout);
            prompt = Regex.Replace(prompt, @"!?\[([^\]\r\n]*)\]\([^\r\n)]*\)", "$1", RegexOptions.None, RegexTimeout);
            prompt = Regex.Replace(prompt, @"(?m)^\s{0,3}(?:#{1,6}\s+|>\s*|[-+*]\s+|\d+[.)]\s+)", "", RegexOptions.None, RegexTimeout);
            prompt = Regex.Replace(prompt, @"(`+)(.*?)\1", "$2", RegexOptions.None, RegexTimeout);
            prompt = Regex.Replace(prompt, @"(\*\*|__|~~)(.*?)\1", "$2", RegexOptions.None, RegexTimeout);
            prompt = Regex.Replace(prompt, @"(?<!\w)(\*|_)(.*?)\1(?!\w)", "$2", RegexOptions.None, RegexTimeout);
            prompt = Regex.Replace(prompt, @"\\([\\`*_{}\[\]()#+.!>-])", "$1", RegexOptions.None, RegexTimeout);
            var title = TaskJson.Compact(WebUtility.HtmlDecode(prompt), 8192);
            return title.Length == 0 ? null : title;
        }
        catch (RegexMatchTimeoutException) { return null; }
    }

    private static JsonObject? FirstTurn(JsonObject state)
    {
        if (state["turnHistory"] is JsonObject historyState && historyState["kind"].Text() == "canonical")
        {
            if (historyState["history"] is not JsonObject history || history["islands"].Objects().FirstOrDefault() is not { } island
                || island["olderBoundary"] is not JsonObject boundary || boundary["status"].Text() != "exhausted"
                || island["entries"].Objects().FirstOrDefault()?["value"].Text() is not { } key) return null;
            return (history["entitiesByKey"] as JsonObject)?[key] as JsonObject;
        }
        if (state["turnsPagination"] is JsonObject pagination && pagination["hasLoadedOldest"] is JsonValue value
            && value.TryGetValue<bool>(out var loaded) && !loaded) return null;
        return state["turns"].Objects().FirstOrDefault();
    }
}
